sway-lib-std/src/storage/storage_api.sw

```diff
@@ -137,11 +133,54 @@ pub fn read<T>(slot: b256, offset: u64) -> Option<T> {
 /// }
 /// ```
 #[storage(write)]
-pub fn clear<T>(slot: b256) -> bool {
-    // Get the number of storage slots needed based on the size of `T` as the ceiling of 
-    // `__size_of::<T>() / 32`
-    let number_of_slots = (__size_of::<T>() + 31) >> 5;
+pub fn clear<T>(slot: b256, offset: u64) -> bool {
+    if __size_of::<T>() == 0 {
+        return true;
+    }
+
+    // Determine how many slots and where the value is to be cleared.
+    let (offset_slot, number_of_slots, _place_in_slot) = slot_calculator::<T>(slot, offset);
 
     // Clear `number_of_slots * 32` bytes starting at storage slot `slot`.
-    __state_clear(slot, number_of_slots)
+    __state_clear(offset_slot, number_of_slots)
+}
+
+/// Given a slot, offset, and type this function determines where something should be stored.
+///
+/// # Arguments
+///
+/// * `slot`: [b256] - The starting address at which something should be stored.
+/// * `offset`: [u64] - The offset from `slot` to store the value.
+///
+/// # Returns
+///
+/// [b256] - The calculated offset slot to store the value.
+/// [u64] - The number of slots the value will occupy in storage.
+/// [u64] - The word in the slot where the value will start.
+fn slot_calculator<T>(slot: b256, offset: u64) -> (b256, u64, u64) {
+    let size_of_t = __size_of::<T>();
+
+    // Get the last storage slot needed based on the size of `T`.
+    // ((offset * bytes_in_word) + bytes + (bytes_in_slot - 1)) >> align_to_slot = last slot
+    let last_slot = ((offset * 8) + size_of_t + 31) >> 5;
+
+    // Where in the storage slot to align `T` in order to pack word-aligned.
+    // offset % number_words_in_slot = word_place_in_slot
+    let place_in_slot = offset % 4;
```

Comment:
Is the division by 4 because we can pack 4 `u64`s into a `b256`?
So we align it to one of the ends (I assume the left / most significant)?

Reply:
1 word = 8 bytes
quad word = 32 bytes
number of words in slot = 4

A storage slot is of size quad word and all storage read/writes are of size quad words as per the `__state_store_quad()` intrinsic function. 

The offset is the offset of words. We want to know which location in the quad word the offset is pointing to.